File: NeuralNetNative/EliminationRegularization.cpp
#define NEURALNETNATIVEAPI
#include <mathimf.h>
#include "EliminationRegularization.h"

namespace NeuralNetNative {
	EliminationRegularization::EliminationRegularization(float regularizationFactor, float alpha) : Regularization(regularizationFactor) {
		_sqrAlpha = alpha*alpha;
	}

	float EliminationRegularization::GetDerivative(float value) {
		float tmp = _sqrAlpha + value*value;
		return 2.0f*RegularizationFactor*_sqrAlpha*value/(tmp*tmp);
	}
}

File: NeuralNet/Regularization/EliminationRegularization.cs
using System;

namespace NeuralNet {
	public sealed class EliminationRegularization : Regularization {
		private readonly float _sqrAlpha;

		public EliminationRegularization(float regularizationFactor, float alpha) : base(regularizationFactor) {
			_sqrAlpha = alpha*alpha;
		}

		public override float GetDerivative(float value) {
			var tmp = _sqrAlpha + value*value;
			return 2.0f*RegularizationFactor*_sqrAlpha*value/(tmp*tmp);
		}

		public float Alpha {
			get { return (float) Math.Sqrt(_sqrAlpha); }
		}
	}
}